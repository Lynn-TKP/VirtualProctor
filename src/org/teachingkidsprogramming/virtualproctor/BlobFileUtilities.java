package org.teachingkidsprogramming.virtualproctor;

import java.util.ArrayList;
import java.util.List;

import org.teachingkidsprogramming.virtualproctor.tests.BlobFile;

import com.spun.util.ObjectUtils;

public class BlobFileUtilities
{

  public static BlobFile[] latestFilesOnly(List<BlobFile> foundFiles)
  {
    ArrayList<BlobFile> keptFiles = new ArrayList<BlobFile>();
    for (BlobFile file : foundFiles)
    {
      boolean latest = true;
      for (BlobFile sub : foundFiles)
      {
        if (ObjectUtils.isEqual(sub.getFileName(), file.getFileName()))
        {
          if (sub.getCreation().after(file.getCreation()))
          {
            latest = false;
          }
        }
      }
      if (latest)
      {
        keptFiles.add(file);
      }
    }
    return keptFiles.toArray(new BlobFile[keptFiles.size()]);
  }
}
