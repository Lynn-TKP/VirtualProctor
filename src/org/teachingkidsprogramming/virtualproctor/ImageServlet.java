package org.teachingkidsprogramming.virtualproctor;

import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import com.google.appengine.api.blobstore.BlobKey;
import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
import com.spun.util.servlets.BasicServlet;

public class ImageServlet extends HttpServlet
{
  @Override
  protected void doGet(HttpServletRequest req, HttpServletResponse res) throws ServletException, IOException
  {
    BlobKey blobKey = new BlobKey(BasicServlet.loadNullableString(req, "fileName"));
    BlobstoreServiceFactory.getBlobstoreService().serve(blobKey, res);
  }
}
