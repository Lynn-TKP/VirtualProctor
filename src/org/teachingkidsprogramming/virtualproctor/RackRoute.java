package org.teachingkidsprogramming.virtualproctor;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

import org.jrack.JRack;
import org.jrack.RackResponse;
import org.jrack.utils.RackEnviromentHelper;

import com.spun.util.MySystem;

public class RackRoute implements JRack
{
  private final JRack        defaultRack;
  private Map<String, JRack> routes = new HashMap<String, JRack>();
  public RackRoute(JRack defaultRack)
  {
    this.defaultRack = defaultRack;
  }
  @Override
  public RackResponse call(Map<String, Object> environment) throws Exception
  {
    String key = RackEnviromentHelper.getPathInfo(environment);
    MySystem.variable("path", key);
    JRack rack = routes.get(key);
    rack = rack == null ? defaultRack : rack;
    return rack.call(environment);
  }
  public RackRoute addRoute(String path, JRack rack)
  {
    routes.put(path, rack);
    return this;
  }
  public RackRoute addRoutesFor(Class<? extends JRack>... classes)
  {
    for (Class<? extends JRack> rackClass : classes)
    {
      try
      {
        Method callMethod = rackClass.getMethod("call", new Class[]{Map.class});
        if (callMethod.isAnnotationPresent(Route.class))
        {
          addRoute(callMethod.getAnnotation(Route.class).value(), rackClass.newInstance());
        }
      }
      catch (Exception e)
      {}
    }
    return this;
  }
}
