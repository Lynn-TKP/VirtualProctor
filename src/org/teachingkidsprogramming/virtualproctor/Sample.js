function loadData() {
	var image = 'http://virtualproctor-tkp.appspot.com/Image?fileName=AMIfv97AeDHZ2uo5BmX_DwelGJAp8XWwJMsXq-US2F5pTSKpbfwCbBznV5cYH_m8eiC3uFE4ZP_B8nbNxrJAeDp1MUHyYPjDy-J0BZhabUlF7X6DO7wwh8a0zPlsGlF5NKNzMNYotXET2o2OG95iEzSzouA3y8tQFmoyBYzXfyEODMaJwQWUgik';

	processImages({
		images : [ {
			fileName : 'lynn',
			imageUrl : image
		}, {
			fileName : 'llewellyn',
			imageUrl : image
		}, {
			fileName : 'person3',
			imageUrl : image
		}, {
			fileName : 'person4',
			imageUrl : image
		}, {
			fileName : 'person5',
			imageUrl : image
		}, {
			fileName : 'person6',
			imageUrl : image
		}, {
			fileName : 'person7',
			imageUrl : image
		}, {
			fileName : 'person8',
			imageUrl : image
		}, {
			fileName : 'person9',
			imageUrl : image
		}, {
			fileName : 'person10',
			imageUrl : image
		} ]
	});
}
$(document).ready(loadData);
