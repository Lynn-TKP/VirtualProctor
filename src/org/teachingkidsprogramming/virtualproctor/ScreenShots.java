package org.teachingkidsprogramming.virtualproctor;

import java.util.HashMap;
import java.util.Map;

import org.apache.sling.commons.json.JSONException;
import org.apache.sling.commons.json.JSONObject;
import org.jrack.JRack;
import org.jrack.RackResponse;
import org.jrack.RackResponseUtils;
import org.teachingkidsprogramming.virtualproctor.tests.BlobFile;

import com.spun.util.ObjectUtils;

public class ScreenShots implements JRack
{
  @Override
  @Route("/ScreenShots")
  public RackResponse call(Map<String, Object> environment) throws Exception
  {
    String json = getJsonForImages(VirtualProctorViewerRack.getAllImagesFromCache());
    return standardJson(json);
  }
  public String getJsonForImages(BlobFile[] files)
  {
    try
    {
      JSONObject array = new JSONObject();
      for (int i = 0; i < files.length; i++)
      {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("fileName", files[i].getFileName());
        jsonObject.put("imageUrl", "/Image?fileName=" + files[i].getKey());
        array.put("" + i, jsonObject);
      }
      array.put("length", files.length);
      JSONObject all = new JSONObject();
      all.put("images", array);
      return all.toString();
    }
    catch (JSONException e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  public static RackResponse standardJson(String html)
  {
    return new RackResponse(RackResponseUtils.ReturnCode.OK, getStandardJsonHeader(), html);
  }
  public static Map<String, String> getStandardJsonHeader()
  {
    Map<String, String> header = new HashMap<String, String>();
    header.put(RackResponseUtils.CONTENT_TYPE, "application/json");
    return header;
  }
}
