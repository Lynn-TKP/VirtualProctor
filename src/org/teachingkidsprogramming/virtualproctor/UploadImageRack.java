package org.teachingkidsprogramming.virtualproctor;

import java.io.InputStream;
import java.nio.channels.Channels;
import java.util.Date;
import java.util.Hashtable;
import java.util.Map;

import javax.servlet.ServletInputStream;
import javax.servlet.http.HttpServletRequest;

import org.jrack.JRack;
import org.jrack.RackEnvironment;
import org.jrack.RackResponse;
import org.jrack.RackResponseUtils;
import org.teachingkidsprogramming.virtualproctor.tests.BlobFile;

import com.google.appengine.api.files.AppEngineFile;
import com.google.appengine.api.files.FileService;
import com.google.appengine.api.files.FileServiceFactory;
import com.google.appengine.api.files.FileWriteChannel;
import com.spun.util.io.FileUtils;

public class UploadImageRack implements JRack
{
  public static Hashtable<String, BlobFile> uploadedFiles = new Hashtable<String, BlobFile>();
  @Override
  public RackResponse call(Map<String, Object> environment) throws Exception
  {
    HttpServletRequest request = (HttpServletRequest) environment.get("HttpServletRequest");
    if (request.getContentLength() > 1)
    {
      ServletInputStream image = request.getInputStream();
      String query = (String) environment.get(RackEnvironment.QUERY_STRING);
      String fileName = query.split("=")[1];
      writeToFile(fileName, image);
    }
    return RackResponseUtils.standardHtml("uploaded");
  }
  public static void writeToFile(String fileName, InputStream in) throws Exception
  {
    FileService fileService = FileServiceFactory.getFileService();
    AppEngineFile file = fileService.createNewBlobFile("image/png", fileName);
    FileWriteChannel channel = fileService.openWriteChannel(file, true);
    FileUtils.copyStream(in, Channels.newOutputStream(channel));
    channel.closeFinally();
    BlobFile blobFile = new BlobFile();
    blobFile.setKey(fileService.getBlobKey(file).getKeyString());
    blobFile.setCreation(new Date());
    blobFile.setFileName(fileName);
    uploadedFiles.put(fileName, blobFile);
  }
}
