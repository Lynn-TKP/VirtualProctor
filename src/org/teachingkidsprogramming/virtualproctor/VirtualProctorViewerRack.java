package org.teachingkidsprogramming.virtualproctor;

import java.util.Date;
import java.util.List;
import java.util.Map;

import org.jrack.JRack;
import org.jrack.RackResponse;
import org.jrack.RackResponseUtils;
import org.teachingkidsprogramming.virtualproctor.tests.BlobFile;

import com.google.appengine.api.blobstore.BlobInfoFactory;
import com.google.appengine.api.datastore.DatastoreService;
import com.google.appengine.api.datastore.DatastoreServiceFactory;
import com.google.appengine.api.datastore.FetchOptions;
import com.google.appengine.api.datastore.Query;
import com.google.appengine.api.datastore.Query.FilterOperator;
import com.spun.util.StringUtils;
import com.spun.util.io.FileUtils;
import com.spun.util.velocity.ContextAware;
import com.spun.util.velocity.VelocityParser;

public class VirtualProctorViewerRack implements JRack
{
  private static Date lastTimeChecked = new Date();
  @Override
  @Route("/")
  public RackResponse call(Map<String, Object> environment) throws Exception
  {
    String html = FileUtils.readFromClassPath(getClass(), "virtual_proctor.html");
    html = StringUtils.stripWhiteSpace(html);
    return RackResponseUtils.standardHtml(html);
  }
  private BlobFile[] getAllImagesFromCache()
  {
    removeOldFiles();
    return UploadImageRack.uploadedFiles.values().toArray(new BlobFile[0]);
  }
  public void removeOldFiles()
  {
    Date tenMinutesAgo = tenMinutesAgo();
    if (lastTimeChecked.before(tenMinutesAgo))
    {
      lastTimeChecked = new Date();
      for (String k : UploadImageRack.uploadedFiles.keySet())
      {
        if (UploadImageRack.uploadedFiles.get(k).getCreation().before(tenMinutesAgo))
        {
          UploadImageRack.uploadedFiles.remove(k);
        }
      }
    }
  }
  public static Date tenMinutesAgo()
  {
    Date tenMinutesAgo = new Date(System.currentTimeMillis() - (10 * 60 * 1000));
    return tenMinutesAgo;
  }
  public RackResponse call(BlobFile[] files)
  {
    String html = VelocityParser.parseFromClassPath(getClass(), "virtual_proctor.html",
        new ContextAware.ContextAwareMap("files", files));
    html = StringUtils.stripWhiteSpace(html);
    return RackResponseUtils.standardHtml(html);
  }
  public static BlobFile[] getAllImages()
  {
    Date tenMinutesAgo = tenMinutesAgo();
    Query query = getFindFilesQuery().addFilter(BlobInfoFactory.CREATION, FilterOperator.GREATER_THAN,
        tenMinutesAgo);
    return getImagesForQuery(query);
  }
  public static BlobFile[] getImagesForQuery(Query query)
  {
    DatastoreService service = DatastoreServiceFactory.getDatastoreService();
    List<BlobFile> foundFiles = BlobFile.wrapAll(service.prepare(query)
        .asList(FetchOptions.Builder.withDefaults()));
    return BlobFileUtilities.latestFilesOnly(foundFiles);
  }
  public static Query getFindFilesQuery()
  {
    return new com.google.appengine.api.datastore.Query("__BlobInfo__");
  }
}
