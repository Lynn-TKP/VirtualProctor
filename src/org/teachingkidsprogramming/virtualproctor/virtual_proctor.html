<html>
	<head>
		<script src="http://code.jquery.com/jquery-1.9.1.js">
		</script>
		<script src="http://code.jquery.com/ui/1.10.3/jquery-ui.js">
		</script>

 <style>
  .div { width: 150px; height: 150px; padding: 0.5em; }
  </style>
		<script>
  var data = {"dateTime":"01/01/01","images":{"0":{"fileName":"llewellyns computer","imageUrl":"http://docs.python.org/3/_images/turtle-star.png"},"1":{"fileName":"lynns computer","imageUrl":"http://nebomusic.net/jesRainbowExample.png"},"length":2}};

$( document ).ready(function() {
  for(var i = 0; i < data.images.length; i++)
  {
    var image = data.images[i];
	var html = '<div id="'+image.fileName+'" class="ui-widget-content" ><img src="' + image.imageUrl + '"><p contenteditable="true">' + image.fileName +'</p></div> &nbsp;' ;
	$( ".inner" ).append(html );
   } 
    $("div").draggable();
});
    </script>
	</head>
	<body class="inner">
        &nbsp;
		</body>
	</html>