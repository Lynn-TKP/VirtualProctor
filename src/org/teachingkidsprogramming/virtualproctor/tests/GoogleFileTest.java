package org.teachingkidsprogramming.virtualproctor.tests;

import java.util.List;

import junit.framework.TestCase;

import org.approvaltests.Approvals;
import org.teachingkidsprogramming.virtualproctor.UploadImageRack;
import org.teachingkidsprogramming.virtualproctor.VirtualProctorViewerRack;

import com.google.appengine.api.datastore.DatastoreService;
import com.google.appengine.api.datastore.DatastoreServiceFactory;
import com.google.appengine.api.datastore.Entity;
import com.google.appengine.api.datastore.FetchOptions;
import com.google.appengine.tools.development.testing.LocalBlobstoreServiceTestConfig;
import com.google.appengine.tools.development.testing.LocalServiceTestHelper;

public class GoogleFileTest extends TestCase
{
  public void testGetFiles() throws Exception
  {
    LocalServiceTestHelper helper = new LocalServiceTestHelper(new LocalBlobstoreServiceTestConfig());
    helper.setUp();
    UploadImageRack.writeToFile("1.png", GoogleFileTest.class.getResourceAsStream("testImage.jpg"));
    UploadImageRack.writeToFile("2.png", GoogleFileTest.class.getResourceAsStream("testImage.jpg"));
    UploadImageRack.writeToFile("3.png", GoogleFileTest.class.getResourceAsStream("testImage.jpg"));
    Approvals.verifyAll("image",
        VirtualProctorViewerRack.getImagesForQuery(VirtualProctorViewerRack.getFindFilesQuery()));
  }
  public void testFileInfo() throws Exception
  {
    LocalServiceTestHelper helper = new LocalServiceTestHelper(new LocalBlobstoreServiceTestConfig());
    helper.setUp();
    UploadImageRack.writeToFile("1.png", GoogleFileTest.class.getResourceAsStream("testImage.jpg"));
    DatastoreService service = DatastoreServiceFactory.getDatastoreService();
    List<Entity> foundFiles = service.prepare(VirtualProctorViewerRack.getFindFilesQuery()).asList(
        FetchOptions.Builder.withDefaults());
    BlobFile wrap = BlobFile.wrap(foundFiles.get(0));
    //BlobInfo wrap2 = new BlobInfoFactory().createBlobInfo(foundFiles.get(0));
    System.out.println(wrap.getKey());
    Approvals.verify(wrap.toString());
  }
  public void testname() throws Exception
  {
    String expected = "samantha";
    assertEquals(expected, "sam");
  }
}
