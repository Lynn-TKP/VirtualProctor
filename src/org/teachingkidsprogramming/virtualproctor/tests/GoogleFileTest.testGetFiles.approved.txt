image[0] = BlobFile [
  key=VQ-1AIPWbhWHRQtC333ojA
  creation=3/19/12 2:29 AM
  creationHandle=writable:GIJAH3SIGe2h3h_ngmJe4Q
  fileName=3.png
  contentType=image/png
  size=1,228
]
image[1] = BlobFile [
  key=Vez9-n9fWBwHVWmk4dgxbw
  creation=3/19/12 2:29 AM
  creationHandle=writable:FFxtvOj0yoq6XfcqLaIWjQ
  fileName=2.png
  contentType=image/png
  size=1,228
]
image[2] = BlobFile [
  key=qFSpny_KUHbRNZI9238flQ
  creation=3/19/12 2:29 AM
  creationHandle=writable:b3LVLMQQmETbRmaLi0NkJw
  fileName=1.png
  contentType=image/png
  size=1,228
]
